src/tests/unittests/README.md

```diff
@@ -0,0 +1,171 @@
+# Unit testing with cmocka
+
+This folder contains unit tests ([https://en.wikipedia.org/wiki/Unit_testing](URL)) for different
+code units of darktable. The tests are written with the help of cmocka ([https://cmocka.org](URL)).
+This is a unit testing framework designed for C with the support of function mocking. By coincident
+the maintainer of cmocka is also a contributor to darktable.
+
+This README explains how unit tests are designed and how a developer can add its own tests.
+
+
+## How to build and run tests
+
+Compiling and running unit tests is very easy. Each unit test suite `test_<suite-name>.c` is built
+as own executable. It can be run as a single program (getting all debug output) or together with
+all other tests (getting just a summary).
+
+To build unit tests, cmocka must be installed (on Ubuntu 18.04: "apt install libcmocka-dev"). Then,
+build darktable with the option `-DBUILD_TESTING=ON`. This tells cmake to build all tests together
+with the darktable binary. Once the binary is built, you can also just type `make
+test_<suite-name>` in the build folder. This will compile just the specified test suite (very
+useful while developing tests since it is much faster than re-building the whole source tree).
+
+To run all tests together, type `make test` in the build folder. This will kick ctest to run all
+tests and provide a summary output. To run a single test suite and get all debug output, you can
+run the binary directly from the build folder, e.g: `./src/tests/unittests/test_sample`.
+
+## Design rules for unit tests
+
+Good code quality requires some general rules in order to harmonize the developers mindset.
+
+### Editor settings
+* Spaces instead of tabs
+* Indentation width: 2 spaces
+* Max line length: 99 chars (allowing single lines to exceed) [there seems to be no consent in the
```

Comment:
Should be 80.

Why do projects use normally 4 spaces and 80 chars for max lines?

There are basically two reasons:

* The first one is to prevent that you nest (loops, if-clauses, switch, etc) too much. If you hit the line limit you should create a function!
* Split windows (at least I use them a lot)

Reply:
Fully agree - but I also vote for consistency. All code I've seen so far in DT has an indentation of 2 spaces and a mixed value of max line length.

I asked in the IRC group about max line length and got very different values. But what I felt was that it doesn't really matter - so I decided to go with 99 (or 100) since that is what fits nicely into 3 splits on a QHD display (and probably 2 splits on HD). 80 is in my eyes generally too short when you want to have reasonable and intuitive variable and method names.

Reply:
[This](https://github.com/darktable-org/darktable/blob/cb5413a59faa8192185a1b339e582231b0789342/.clang-format#L17) needs fixing.

Reply:
Checked the IRC log quickly, it was houz who stated 120. So, we already have 80 and 120 :-) I don't want to have a long discussion here about line length since that's going to be endless (there's no hard argument pro/contra any certain number). The design rules presented here are only valid for unit tests anyway.

Can you agree on a "soft" requirement of 100, as a compromise between 80 and 120? The target is readability and this does not depend so much on line length.

Reply:
Well I agree wit @cryptomilk the line limit should be 80 no more. That the only way to have easy to read side-by-side diff for example. The news paper have  small line to help reading, at the end of a long line it is far more complex to go to next one. And BTW, better ask the developers there as message to IRC where most of them are not!

I also plan on removing the script to automatically reformat the code. I have made a lot of effort to make the SQL statement readable and don't want to loose this because of a stupid script.

Reply:
Looks like I was wrong, clang-format keeps the strings as-is! So the completely obscure indentation of SQL statements were made by devs !

Anyway, line length of 80 is still a good goal to me.

Reply:
I was not completely wrong, if the line is too wide and need reindentig we get this:

```
+  DT_DEBUG_SQLITE3_PREPARE_V2(
+      dt_database_get(darktable.db),
+      "INSERT INTO main.images"
+      "  (id, group_id, film_id, width, height, filename, maker, model, lens, "
+      "exposure,"
+      "   aperture, iso, focal_length, focus_distance, datetime_taken, flags,"
+      "   output_width, output_height, crop, raw_parameters, "
+      "raw_denoise_threshold,"
+      "   raw_auto_bright_threshold, raw_black, raw_maximum,"
+      "   caption, description, license, sha1sum, orientation, histogram, "
+      "lightmap,"
+      "   longitude, latitude, altitude, color_matrix, colorspace, version, "
+      "max_version, history_end,"
+      "   position, aspect_ratio)"
+      " SELECT NULL, group_id, film_id, width, height, filename, maker, model, "
+      "lens,"
+      "       exposure, aperture, iso, focal_length, focus_distance, "
+      "datetime_taken,"
+      "       flags, width, height, crop, raw_parameters, "
+      "raw_denoise_threshold,"
+      "       raw_auto_bright_threshold, raw_black, raw_maximum,"
+      "       caption, description, license, sha1sum, orientation, histogram, "
+      "lightmap,"
+      "       longitude, latitude, altitude, color_matrix, colorspace, NULL, "
+      "NULL, 0, ?1, aspect_ratio"
+      " FROM main.images WHERE id = ?2",
+      -1, &stmt, NULL);
```

Reply:
Well, you wanted line length 80, you got it I guess.

I would vote for 100-120, 80 is too little.

Reply:
@TurboGit thanks for your answer. 80 is fine for me. I will update the PR.